File: tools/test_conditions.mjs
/* 条件チップが、表示している件数どおりに絞り込めるかを実ブラウザで確かめる。
 *
 *   cd web && python3 -m http.server 8144 &
 *   node tools/test_conditions.mjs http://localhost:8144
 *
 * ■ なぜブラウザ側にも要るか
 * CONDITIONS は server.py と web/assets/app.js の両方にある。
 * **本番は静的配信なので、実際に絞り込んでいるのは app.js のほう。**
 * server.py だけ直しても本番は0件のまま、という直し漏れが起きる。
 * 件数は数え直さず「チップに出ている数」と「絞った結果の件数」を
 * 突き合わせる ―― 期待値をここに書くと tools/test_conditions.py と
 * 二重管理になり、必ず片方が古くなる。
 */
import { chromium } from "playwright";

const base = process.argv[2] || "http://localhost:8144";
const b = await chromium.launch();
const p = await b.newPage({ viewport: { width: 1280, height: 1400 } });
const errs = [];
p.on("pageerror", e => errs.push(String(e)));

let n = 0;
const fails = [];
const check = (cond, msg) => { n++; if (!cond) fails.push(msg); };

// 開屏の問診に邪魔されないよう、済んだことにしてから開く（test_favorite.mjs / test_sort.mjs と同じ）。
await p.addInitScript(() => {
  try { localStorage.setItem("rk_onboarded", "1"); } catch (e) {}
  try { sessionStorage.setItem("rk_splash_seen", "1"); } catch (e) {}
});

await p.goto(base, { waitUntil: "networkidle" });
await p.waitForSelector("#list > .card");

/* 既定は学年も学期も「すべて」。ここが変わると件数が全部変わるので先に確かめる。 */
check(await p.$eval("#years .chip.on", e => e.textContent.trim()) === "すべて",
      "学年の既定が「すべて」ではない");
check(await p.$eval("#sems .chip.on", e => e.textContent.trim()) === "すべて",
      "学期の既定が「すべて」ではない");

const chips = async () => p.$$eval("#conds .chip, #trust .chip", cs =>
  cs.map(c => ({
    name: c.textContent.replace(/\d+$/, "").trim(),
    count: +(c.querySelector(".n")?.textContent ?? 0),
  })));

const before = await chips();
check(before.length >= 6, `条件チップが少なすぎる（${before.length}個）`);

/* 「0件のチップを作らない」は**全所属ぶんのデータがあるときだけ**の話。
   server.py を data/courses.json（共通教育1,112件）だけで動かすと、
   集中講義191件は全部が学部の専門科目なので **正しく0件**になる。
   データ不足を不具合として落とさないよう、母数で判定を分ける。 */
const total = +(await p.$eval("#count", e => e.textContent)).replace(/[^\d]/g, "");
const full = total >= 7000;
console.log(`  母数 ${total}件（${full ? "全所属" : "一部のみ ― 0件の判定は飛ばす"}）`);

for (const { name, count } of before) {
  // 「押せるのに1件も出ない」チップを作らない。2026-08-26 まで
  // 出席なし・レポートのみ・集中講義 の3つがこの状態だった。
  if (full) check(count > 0, `「${name}」が0件（押せるのに1件も出ない）`);
  if (count === 0) continue;

  // チップの数字どおりに絞れること。ここが食い違うと、
  // 数えている実装と絞っている実装がずれている。
  const btn = p.locator(`#conds .chip, #trust .chip`).filter({ hasText: name }).first();
  await btn.click();
  await p.waitForTimeout(400);
  const got = +(await p.$eval("#count", e => e.textContent)).replace(/[^\d]/g, "");
  check(got === count, `「${name}」チップは ${count}件 と出ているのに、押すと ${got}件`);
  await btn.click();                    // 外して次へ
  await p.waitForTimeout(300);
}

check(!errs.length, "コンソールに例外: " + errs.join(" | "));

console.log(`  条件チップ ${before.length}個: ` +
            before.map(c => `${c.name} ${c.count}`).join(" / "));
console.log(`  通過 ${n - fails.length} 件 / ${n} 件`);
fails.forEach(f => console.log("  NG  " + f));
console.log(fails.length ? "NG" : "OK");
await b.close();
process.exit(fails.length ? 1 : 0);

File: tools/test_conditions.py
"""条件チップ（出席なし・レポートのみ…）の判定のテスト。ネットワークには出ない。

■ なぜこれが要るか
2026-08-26 まで「出席なし」「レポートのみ」「集中講義」の3つが、
**全7,877件に対して常に0件**だった。押せるのに1件も出ない状態が、
誰にも気付かれずに本番へ出ていた。件数を機械に数えさせておけば、
同じことが二度起きても落ちる。

  python3 tools/test_conditions.py
"""
import contextlib
import io
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
# server.py は読み込むだけで科目データを開き、ダミーなら警告を出す。
# ここで欲しいのは CONDITIONS だけなので、その出力は飲み込む。
with contextlib.redirect_stdout(io.StringIO()):
    import server

ROOT = Path(__file__).resolve().parent.parent
BUILT = ROOT / "web" / "data" / "courses.built.json"

fails = []
n = 0


def check(cond, msg):
    global n
    n += 1
    if not cond:
        fails.append(msg)


check(BUILT.is_file(), "web/data/courses.built.json が無い")
if BUILT.is_file():
    courses = json.loads(BUILT.read_text(encoding="utf-8"))["courses"]
    C = server.CONDITIONS
    hit = {name: [c for c in courses if f(c)] for name, f in C.items()}

    # ① どのチップも「押せるのに1件も出ない」になっていないこと。
    #    口コミあり だけは、口コミが0件の環境では 0 になりうるので別扱い。
    for name, rows in hit.items():
        if name == "口コミあり":
            continue
        check(len(rows) > 0, f"「{name}」が0件（押せるのに1件も出ない）")

    # ② 件数。データを焼き直して動いたら、意図した変化か確かめる。
    expect = {"出席なし": 1536, "レポートのみ": 482, "集中講義": 193,
              "持ち込み可": 17, "1限以外": 6979, "小テストなし": 6453}
    for name, want in expect.items():
        got = len(hit[name])
        check(got == want, f"「{name}」の件数が {want} → {got} に変わった")

    # ③ 「レポートのみ」は、毎回小テストが無ければ必ず「出席なし」でもある
    #    （成績の内訳に出席が入っていないため）。
    #    例外は**毎回小テストがある科目だけ**。成績には効かないが毎週の拘束は
    #    あるので「出席なし」からは外している。いま該当するのは
    #    083119 海事政策論 の1件（レポート100%＋毎回小テスト）。
    #    片方だけ直すと崩れる関係なので、機械に持たせておく。
    stray = [c["id"] for c in hit["レポートのみ"]
             if not server.CONDITIONS["出席なし"](c) and not c.get("weekly_quiz")]
    check(not stray,
          f"「レポートのみ」なのに「出席なし」でない（毎回小テストも無い）: {stray[:5]}")

    # ④ 内訳が最後まで分かっていない科目を、この2つに含めないこと。
    #    未分類の残りに出席や試験が隠れている可能性がある。
    for name in ("出席なし", "レポートのみ"):
        bad = [c["id"] for c in hit[name]
               if not c.get("eval_ratio") or c.get("eval_unclassified")]
        check(not bad, f"「{name}」に内訳の不完全な科目が入っている: {bad[:5]}")

    # ⑤ 0% はキーを落とす形で表れる（scrape/parse.py:152）。
    #    もし将来 0 を持つようになったら、判定を書き直す合図。
    zero = [c["id"] for c in courses
            for k in ("attendance", "exam", "report")
            if (c.get("eval_ratio") or {}).get(k) == 0]
    check(not zero,
          f"eval_ratio に 0% のキーが現れた（判定の前提が変わった）: {zero[:5]}")

    # ⑥ 「集中講義」は開講区分（term）で見る。class_format には該当値が無い。
    check(not any(c.get("class_format") == "集中講義" for c in courses),
          "class_format に '集中講義' が現れた（判定を見直すこと）")

print(f"  通過 {n - len(fails)} 件 / {n} 件")
for f in fails:
    print(f"  NG  {f}")
print("OK" if not fails else "NG")
sys.exit(1 if fails else 0)
